streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSegmentedBytesStore.java

```diff
@@ -0,0 +1,337 @@
+                                          final String metricsScope,
+                                          final long retention,
+                                          final long segmentInterval,
+                                          final boolean withIndex) {
+        super(name, metricsScope, new TimeFirstWindowKeySchema(),
+            Optional.ofNullable(withIndex ? new KeyFirstWindowKeySchema() : null),
+            new KeyValueSegments(name, metricsScope, retention, segmentInterval));
+    }
+
+    public void put(final Bytes key, final long timestamp, final int seqnum, final byte[] value) {
+        final Bytes baseKey = TimeFirstWindowKeySchema.toStoreKeyBinary(key, timestamp, seqnum);
+        put(baseKey, value);
+    }
+
+    byte[] fetch(final Bytes key, final long timestamp, final int seqnum) {
+        return get(TimeFirstWindowKeySchema.toStoreKeyBinary(key, timestamp, seqnum));
+    }
+
+    @Override
+    protected KeyValue<Bytes, byte[]> getIndexKeyValue(final Bytes baseKey, final byte[] baseValue) {
+        final byte[] key = TimeFirstWindowKeySchema.extractStoreKeyBytes(baseKey.get());
+        final long timestamp = TimeFirstWindowKeySchema.extractStoreTimestamp(baseKey.get());
+        final int seqnum = TimeFirstWindowKeySchema.extractStoreSequence(baseKey.get());
+
+        return KeyValue.pair(KeyFirstWindowKeySchema.toStoreKeyBinary(key, timestamp, seqnum), new byte[0]);
+    }
+
+    @Override
+    Map<KeyValueSegment, WriteBatch> getWriteBatches(
+        final Collection<ConsumerRecord<byte[], byte[]>> records) {
+        // advance stream time to the max timestamp in the batch
+        for (final ConsumerRecord<byte[], byte[]> record : records) {
+            final long timestamp = WindowKeySchema.extractStoreTimestamp(record.key());
+            observedStreamTime = Math.max(observedStreamTime, timestamp);
+        }
+
+        final Map<KeyValueSegment, WriteBatch> writeBatchMap = new HashMap<>();
+        for (final ConsumerRecord<byte[], byte[]> record : records) {
+            final long timestamp = WindowKeySchema.extractStoreTimestamp(record.key());
+            final long segmentId = segments.segmentId(timestamp);
+            final KeyValueSegment segment = segments.getOrCreateSegmentIfLive(segmentId, context, observedStreamTime);
+            if (segment != null) {
+                ChangelogRecordDeserializationHelper.applyChecksAndUpdatePosition(
+                    record,
+                    consistencyEnabled,
+                    position
+                );
+                try {
+                    final WriteBatch batch = writeBatchMap.computeIfAbsent(segment, s -> new WriteBatch());
+
+                    // Assuming changelog record is serialized using WindowKeySchema
+                    // from ChangeLoggingTimestampedWindowBytesStore. Reconstruct key/value to restore
+                    final byte[] key = WindowKeySchema.extractStoreKeyBytes(record.key());
```

Comment:
I concern that by stripping off just the first byte for base key and index key, we would be paying too much byte array copies which would be much slower than other stores (for which we just blindly copy-paste the bytes). Maybe we can have some optimization in-place for the first draft? E.g.:

1) for index key, we know we just need to add the first byte and strip the last four byte of seq (? is that right, need to double check).

2) for base key, we can just do a swap-copy from the index key to switch the position of timestamp and key.

Reply:
Yeah. Index could be just adding prefix and base key could be constructed directly from index key.